### PR TITLE
`dev` improvements: Load extensions using the Specifications from Core

### PR DESCRIPTION
**Pull request by isaacroldan**
### WHY are these changes introduced?
As part of the extensions revamp project, we need to validate extensions with the remote extension specifications API.

### WHAT is this pull request doing?
- When loading the app with `loader.ts`, you now need to provide a list of supported specifications.
- For `dev`, fetch all the specs before loading the app, so that the app can only be loaded if all current extensions are supported.
- Other commands that also load the local app but don't need remote specs validation will just use the local specs (like `build`, `env pull`, `env show`, etc...)
- The function `fetchSpecifications` now returns all specs available for the current user/app, overriding local values with remote ones.
- More tests.

### How to test your changes?
- `dev` works for all extensions

### Measuring impact

How do we know this change was effective? Please choose one:

- [ ] n/a - this doesn't need measurement, e.g. a linting rule or a bug-fix
- [ ] Existing analytics will cater for this addition
- [ ] PR includes analytics changes to measure impact

### Checklist

- [ ] I've considered possible cross-platform impacts (Mac, Linux, Windows)
- [ ] I've considered possible [documentation](https://shopify.dev) changes
- [ ] I've made sure that any changes to `dev` or `deploy` have been reflected in the [internal flowchart](https://www.figma.com/file/7vqUp50u6dm48Zfb4JRRn8/CLI3-Internals).

### CODE REVIEW

**Review by pepicrft — COMMENTED** [review]
I tried to tophat the fixture app in the repository and I got the following error:

image

Do I need to be given access to that extension type?

File: packages/app/src/cli/commands/app/env/show.ts [hunk 1]
```diff
@@ -16,8 +17,8 @@ export default class EnvShow extends Command {
   public async run(): Promise<void> {
     const {flags} = await this.parse(EnvShow)
     const directory = flags.path ? path.resolve(flags.path) : process.cwd()
-    const envFile = path.resolve(directory)
-    const app: AppInterface = await loadApp(directory, 'report')
+    const specifications = await allLocalSpecs()
+    const app: AppInterface = await loadApp(directory, specifications, 'report')
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/commands/app/env/show.ts:21]: yep, makes sense 👌

File: packages/app/src/cli/utilities/extensions/fetch-extension-specifications.ts [hunk 2]
```diff
@@ -9,13 +13,20 @@ import {FlattenedRemoteSpecification} from '@shopify/cli-kit/src/api/graphql/ext
...
+ *
  * Will return a merge of the local and remote specs (remote values override local ones)
  * Will only return the specifications that are also defined locally
+ * (Functions are not validated againts remote specs, gated access is defined locally)
+ *
  * @param token - Token to access partners API
  * @returns List of extension specifications
  */
-export async function fetchExtensionSpecifications(token: string, apiKey: string): Promise<GenericSpecification[]> {
+export async function fetchSpecifications(token: string, apiKey: string): Promise<GenericSpecification[]> {
```
[the remark this answers is not in the document]
Comment (isaacroldan) [reply packages/app/src/cli/utilities/extensions/fetch-extension-specifications.ts:29]: not sure we want to introduce the concept of gated in the name of the function 🤔 
I think it's safe to assume that the result will be different depending on the apiKey provided no?

**Comment by isaacroldan** [comment]
For the customer_accounts extension: yes, you need to enable beta access to it. Also for the new ui_extension

**Pull request merged**